TypeSystem/README.md

```diff
@@ -78,16 +87,103 @@ CURRENT_STRUCT(Foo) {
   CURRENT_CONSTRUCTOR(Foo)(uint64_t i) : i(i) {}
 };
 ```
+ 
+### Templates
+
+Along with `CURRENT_STRUCT`, Current type system supports `CURRENT_STRUCT_T`: effectively a `template<typename T> struct ...`. The `CURRENT_CONSTRUCTOR_T` and `CURRENT_DEFAULT_CONSTRUCTOR_T` macros should be used instead of the non-`_T` ones.
+
+### Field Descriptions
+
+When a `CURRENT_STRUCT` defined a `CURRENT_FIELD(foo, ...);`, it may also define `CURRENT_FIELD_DESCRIPTION(foo, "This is the story of foo.");` Like fields, field descriptions can be reflected upon. Field descriptions can also be exported in various formats, most notably C++ itself (as the definition of the C++ struct), and (Markdown)[https://github.com/C5T/Current/blob/master/TypeSystem/Schema/golden/smoke_test_struct.md].
+
+## `Optional`
+
+A special type `Optional<T>` is used to define `T` or `null`. C++ usage of `Optional<T> x;` is:
+* an `if (Exists(x))` condition, and
+* a `T y = Value(x)` retrieval.
+
+This syntax returns an xvalue, which can be captured into a reference, assigned to, etc.
+
+## `Variant`
+
+For polymorphic types, `Variant<TS...>` can be used. C++ usage of `Variant<A, B> x` is:
+* an `if (Exists<A>(x))` condition,
+* a `A a = Value<A>(x)` retrieval, or
+* the `x.Call(visitor)` pattern, where `visitor` is an instance of a struct/class that can be `operator()`-called for any of the underlying types.
+ 
+The former two syntaxes return an xvalue, which can be captured into a reference, assigned to, etc.
+
+The latter syntax is recommended as it's a) more efficient, and b) ensures the compile-type guarantee no inner type is left out.
```

Comment:
Maybe "...recommended **for implementing type dispatching**..." or smth like that?